File: Data Structures & Algorithms/binary-tree-right-side-view/submission-2.js
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     constructor(val = 0, left = null, right = null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    /**
     * @param {TreeNode} root
     * @return {number[]}
     */
    rightSideView(root) {
        if(!root){
            return []
        }
        let currArr = [root]
        let nextArr = []
        let res = []
        console.log(currArr)
        console.log(nextArr)



        while (currArr.length || nextArr.length){
            let node 
            while(currArr.length){
                if(currArr.length == 1){
                    node = currArr.shift()
                    res.push(node.val)
                    if(node.left){
                        nextArr.push(node.left)
                    }
                    if(node.right){
                        nextArr.push(node.right)
                    }
                }
                else{
                    node = currArr.shift()
                     if(node.left){
                        nextArr.push(node.left)
                    }
                    if(node.right){
                        nextArr.push(node.right)
                    }
                }
                console.log(node.val)
            }

            while(nextArr.length){
                if(nextArr.length == 1){
                    node = nextArr.shift()
                    res.push(node.val)
                     if(node.left){
                        currArr.push(node.left)
                    }
                    if(node.right){
                        currArr.push(node.right)
                    }
                }
                else{
                    node = nextArr.shift()
                    if(node.left){
                        currArr.push(node.left)
                    }
                    if(node.right){
                        currArr.push(node.right)
                    }
                }
                 console.log(node.val)
            }
        }
        return res
    }
}

File: Data Structures & Algorithms/products-of-array-discluding-self/submission-2.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        right = [0] * len(nums)
        left = [0] *len(nums)
        ans = [0] *len(nums)

        left[0] = right[len(nums) - 1] = 1
        #left 
        for i in range(1, len(nums)):
            left[i] = left[i - 1] * nums[i - 1]

        #right
        for i in range(len(nums) - 2, -1 ,-1):
            right[i] = right[i + 1] * nums[i + 1]
        
        for i in range(len(nums)):
            ans[i] = left[i] * right[i]
        return ans
